Added validator plugin (to be further modified)

JasonWangCY opened the pull request.

tsengj10 commented:
Adding the new "in progress" label based on the title

tsengj10 commented on snewpdag/plugins/Validator.py:
```diff
@@ -0,0 +1,75 @@
+            return False
+
+    def check_type(self, data): # check that the key corresponed to the correct type
+        return isinstance(data, self.key_type)
```
Does this work? self.key_type will be a string, but I think arg2 needs to be the type/class

JasonWangCY replied:
All the unit tests seem to be working with no errors when I use self.keytype for isinstance(), so I suppose it does work!

JasonWangCY replied:
I think my unit tests work because I didn't treat self.key_type itself as a string, but rather as a type per se.

For instance, when calling the plugin in the **unit test**, instead of writing
     Validator_keytype( key_type = 'str' )         # where the class of key_type itself is a string here,
I wrote 
     Validator_keytype( key_type = str )          # where the class of key_type itself is a type here.

However, I'm not sure if this is legal, but I suppose I can avoid this problem in **the plugin itself** by checking the type conditions one-by-one instead of using an input, e.g.
     if data = 'action'
          return isinstance(data, str)
     elif data = 'dt'
          return isintance(data, float)
     elif
     .
     .
     .

Though I imagine this would be a bit more tedious (I'm not too sure if this is what you meant in the meeting today!)

tsengj10 commented on snewpdag/plugins/Validator.py:
```diff
@@ -0,0 +1,75 @@
+    def check_sorted(self, data):
+        listorder = None
+        if data == sorted(data):
+            listorder = 'ascending'
```
this is going to be pretty slow, since it requires a whole sort operation.  A good sort algorithm will take ~N log N time, others worse.  But here one only needs to check that one element is greater or less than the previous one, and if it's not, flag an error (should take ~N time).

JasonWangCY replied:
Noted, I've removed the sorting functionality from the plugin now, thanks!

tsengj10 commented on snewpdag/tests/test_validator.py:
```diff
@@ -0,0 +1,84 @@
+from snewpdag.plugins import Validator
+
+class TestValidator(unittest.TestCase):
+    
```
good use of unit tests to check the individual parts of the class

tsengj10 commented:
I guess a next step was to write an action method for Validator?

JasonWangCY commented:
I've split the validator into separate plugins, and implemented the action methods. I only included 'alert' so far, but I plan to just copy the same lines for the other actions.

Also, for the checking of the list order, I'm not too sure if any action method is required, since in practice, even if the list is unsorted, the 'alert' should still proceed as normal right?

tsengj10 commented on snewpdag/plugins/Validator_key.py:
```diff
@@ -0,0 +1,31 @@
+            return False
+    
+    def alert(self, data):
+        if self.check_key(data):
```
In other words, `return self.check_key(data)`

tsengj10 commented:
For sorting, perhaps the best response for finding an unsorted list is to sort it.  I still think it's better to do the simple check first, though.  One should also specify the sort order in the configuration (as an argument for __init__), because downstream code will probably need to assume one or the other.

tsengj10 commented on snewpdag/plugins/Validator_listtype.py:
```diff
@@ -0,0 +1,43 @@
+        for x in data:
+            if isinstance(x, self.key_type) == False:
+                remove_element.append(x)
+        if len(remove_element) <= (data_len*0.1):
```
The 0.1 should be a configuration parameter, perhaps with a name like `min_fraction`.  There's also no need to record the removed elements; just count them.

tsengj10 commented:
Also, for checking the type, I'd have written if statements comparing with type names, like 'str', 'int', 'float', 'number' (for int or float), etc.  Trying to interpret the field names would be not only tedious but I suspect impossible, since we're still writing new modules.

tsengj10 commented:
And, finally (!), let's keep the module names using CamelCase convention, i.e., ValidateScalar, ValidateList, ValidateSort, etc.  For all the other names we will tend to use lower-case pythonesque names, like this_is_a_python_name.

tsengj10 commented:
Back to checking type:  if you don't want to write a bunch of if statements, you can also try the formula `type(v).__name__`, which returns the name of the type of the variable v.

tsengj10 commented on snewpdag/plugins/ValidateListType.py:
```diff
@@ -0,0 +1,69 @@
+class ValidateListType(Node):
+    def __init__(self, max_fraction, **kwargs):
+        self.max_fraction = max_fraction
+        self.key_type = kwargs.pop('key_type', None)
```
It seems to me key_type should also be a required argument.  kwargs.pop() is used for optional ones.

tsengj10 commented on snewpdag/plugins/ValidateListType.py:
```diff
@@ -0,0 +1,69 @@
+            logging.error('More than 10% in list are not the desired type of {} and action is consumed'.format(self.key_type))
+            return False
+    
+    def alert(self, data):
```
these methods can be more concise, as you did in ValidateKey

JasonWangCY replied:
I wrote it as 
      if self.check_listtype(data) != False:
            return True
      else:
            return False
because self.check_listtype(data) may return the updated data array itself, instead of a pure True/False boolean, so I'm not sure if there is a cleaner way to do this.

tsengj10 commented on snewpdag/plugins/ValidateSort.py:
```diff
@@ -0,0 +1,41 @@
+
+class ValidateSort(Node):
+    def __init__(self, listorder, **kwargs):
+        self.listorder = listorder
```
listorder should be list_order to be more "pythonesque"

tsengj10 commented:
Nice @JasonWangCY!

For ValidateSort,
1. print() should instead use the logger
2. it seems appropriate to sort the array in case it's found to be out of order. So first one checks (this is the quick operation which hopefully passes most of the time), and, if there are elements out of order, make a copy (the usual payload issue), sort it, and pass the new payload along

tsengj10 commented:
Whoops. Just realized that most of the checks you're doing are on the payload itself, rather than on fields within the payload.

ValidateKey seems to be correct, in that you check that the key exists in the payload.  But the others should allow the user to configure which field (key) to check, check that the field exists in the payload (which does duplicate ValidateKey, but there may be cases for ValidateKey by itself), and then check that the associated value from the payload has the correct characteristics.

JasonWangCY commented:
My intention was that for ValidateKey, I would pass the whole payload called "data" and check if the key exists.
e.g. data = { 'dt' : [0.1, 0.5] }
After checking the key 'dt' does exist, I can then pass it to the other plugins like ValidateSort by directly using data['dt'] as the input. This way, there is no duplication with the ValidateKey plugin, but I suppose I should specify it better in the comments.

tsengj10 commented:
Thanks for the clarification on what you were intending.  But I ValidateKey doesn't call the other plugins per se.  Instead, it checks its payload and, if it passes (by checking that the key exists), it passes the payload on to the next plugin, whatever it is.

The next plugin can assume the key exists, so you don't need to repeat that check.  But you still need to specify the key for which to check the type.  In other words, you still need to tell ValidateSort that it's supposed to look at the dt field in the payload it receives.

(An alternative would be to sub-class ValidateSort from ValidateKey, and have ValidateSort.alert() call super().alert().)

JasonWangCY commented:
I've added an input for all plugins to require in_field, which is the key we are looking at. I've also noticed that I didn't actually change the payload for ValidateListType and ValidateSort, so I've modified those (but I'm not 100% sure if it violates the shallow copy rule you mentioned before!)

tsengj10 commented:
Two last suggestions (I think):
1. change the defaults for on_alert to True, and for the other actions to False.  I think None will be interpreted as False, but it really be a boolean.  Also the usual use case is to check the fields on an alert.
2. ValidateSort has kind of an interesting logic:  if the list is already sorted in one direction, you leave it as it is.  But presumably if the user specifies a list order, it's expected that in the end the list should be sorted in this order?  In other words, if I want the list in ascending order, I want that order regardless of whether the listed had previously been sorted in descending order.

I think you handled the payload change correctly in ValidateListType and ValidateSort, in that you construct a new list and then replace the old list in the dictionary.  This should then be reflected in the downstream payloads, but not those in parallel streams.  (The problem comes when you modify a list in place, because that will be reflected in parallel streams.)

JasonWangCY commented:
Thanks! I've made the suggested changes. I've also changed list_order in ValidateSort to become an optional argument. If no specified list_order is given, the plugin will only check the order of the payload (ascending/descending/None) and return it unmodified, just as an additional feature (though not sure if it's useful!)

tsengj10 commented:
Looks good.  If it's no longer "in progress", please go ahead and remove the label, and I'll flag it for merging soon.
